core/src/main/java/org/infinispan/manager/DefaultCacheManager.java

```diff
@@ -134,7 +136,8 @@
    private final CacheContainerStats stats;
    private final ConfigurationManager configurationManager;
 
-   @GuardedBy("this")
+   private final ReadWriteLock stoppingLock = new ReentrantReadWriteLock();
+   @GuardedBy("stoppingLock")
    private boolean stopping;
```

Comment:
yes, you have to synchronize on something to change the flag (before it was `DefaultCacheManager` instance, which leads to deadlock) and without flag and just blocking on the lock would results into even more deadlocks